pkg/clients/sns/snstopic.go

```diff
@@ -0,0 +1,165 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sns
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/sns"
+
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+
+	"github.com/crossplane/provider-aws/apis/notification/v1alpha1"
+	awsclients "github.com/crossplane/provider-aws/pkg/clients"
+)
+
+const (
+	//SNSTopicNotFound is the error code that is returned if SNS Topic is not present
+	SNSTopicNotFound = "InvalidSNSTopic.NotFound"
+)
+
+// IsTopicNotFound returns true if the error code indicates that the
+// item was not found
+func IsTopicNotFound(err error) bool {
+	if _, ok := err.(*TopicNotFound); ok {
+		return true
+	}
+	return false
+}
+
+// TopicNotFound will be raised when there is no SNSTopic
+type TopicNotFound struct{}
+
+func (err *TopicNotFound) Error() string {
+	return fmt.Sprint(SNSTopicNotFound)
+}
+
+// TopicClient is the external client used for SNSTopic custom resource
+type TopicClient interface {
+	CreateTopicRequest(*sns.CreateTopicInput) sns.CreateTopicRequest
+	ListTopicsRequest(*sns.ListTopicsInput) sns.ListTopicsRequest
+	DeleteTopicRequest(*sns.DeleteTopicInput) sns.DeleteTopicRequest
+	GetTopicAttributesRequest(*sns.GetTopicAttributesInput) sns.GetTopicAttributesRequest
+	SetTopicAttributesRequest(*sns.SetTopicAttributesInput) sns.SetTopicAttributesRequest
+}
+
+// NewTopicClient returns a new client using AWS credentials as JSON encoded data.
+func NewTopicClient(conf *aws.Config) (TopicClient, error) {
+	return sns.New(*conf), nil
+}
+
+// GetSNSTopic returns SNSTopic if present in list of topics
+func GetSNSTopic(res *sns.ListTopicsResponse, cr *v1alpha1.SNSTopic) (sns.Topic, error) {
+
+	for _, topic := range res.Topics {
+		if aws.StringValue(topic.TopicArn) == meta.GetExternalName(cr) {
+			return topic, nil
+		}
+	}
+	return sns.Topic{}, &TopicNotFound{}
+}
+
+// GenerateCreateTopicInput prepares input for CreateTopicRequest
+func GenerateCreateTopicInput(p *v1alpha1.SNSTopicParameters) *sns.CreateTopicInput {
+	input := &sns.CreateTopicInput{
+		Name: &p.Name,
+	}
+
+	if len(p.Tags) != 0 {
+		input.Tags = make([]sns.Tag, len(p.Tags))
+		for i, val := range p.Tags {
+			input.Tags[i] = sns.Tag{
+				Key:   aws.String(val.Key),
+				Value: val.Value,
+			}
+		}
+	}
+
+	return input
+}
+
+// LateInitializeTopic fills the empty fields in *v1alpha1.SNSTopicParameters with the
+// values seen in sns.Topic.
+func LateInitializeTopic(in *v1alpha1.SNSTopicParameters, topic sns.Topic, attrs map[string]string) {
+	in.Name = *awsclients.LateInitializeStringPtr(&in.Name, topic.TopicArn)
```

Comment:
Do we really want to late initialise the name to the ARN? Aren't they different things? Is there ever a case that the name would be empty in our CR? It seems like a required and immutable field.